enhancements/proposed/20190930-addon-host-os-hooks.md

```diff
@@ -0,0 +1,215 @@
+    Instead of asking the developer to provide the minikube ip what if what's app could just identify it
+    automatically. What if I had a whats app business API addon that could be installed and 
+    seemlessly integrated with a host os hook?
+- A desktop UI could be developed for minikube which could be enabled and installed as an addon which would show 
+  the different instances, mounts, etc currently configured in minikube
+- A background service that helps multiple minikube clusters interact with each other can be developed to run on the Host OS
+  This tool could be enabled and started as an addon if these hooks are enabled
+- Tools which need to mount code from the Host OS to the minikube cluster can be developed and enabled through enabling
+  of an addon rather than as a separate step a developer has to do after using minikube. A potential improvement
+  to how skaffold is currently deploying things directly through containers
+
+## Goals
+
+* Make it easy for third party software to be run on the Host OS when a addon is enabled
+* A simple solution is implemented that can achieves the following capabilities:
+  * Plugin owners are responsible for and can easily identify a Host OS
+  * A shell command that is appropriate for that particular host is able to run on the Host OS for the 
+    following events
+    * pre-enable: An addon is about to be enabled
+    * post-enable: An addon has been enabled
+    * pre-disable: An addon is about to be disabled
+    * post-disable: An addon has been disabled
+  * Any program hook can obtain user input like requesting sudo permission to enable and/or start a systemd service
+
+## Non-Goals
+
+* Lifecycle events for addons other than the ones listed above
+* Conditional logic for whether or not to run the addon other than information related to obtaining the Host OS and/or version number
+* Running commands in the background
+* Running commands with administrative privileges eg: sudo without requesting the user to authenticate
+* Changing how a addon is enabled or disabled by default
+
+## Design Details
+
+A configuration file `config.yaml` optionally exists within the `deploy/addons/addonname` directory.
+The configuration file optionally specifies the hooks it wants to run when the addon is enabled or disabled
+Here is an approximate example config: 
+```yaml
+apiVersion: minikube/v1beta
+kind: AddonConfig
+spec:
+  hooks:
+    host:
+      # (optional) Runs just before the addon is enabled
+      preEnable:
+        # A list of potential commands to be run based on the OS and OS version for unix operating systems
+        # Based on the order, the first regex to match the output of the getHostOSInfoCommand will be run
+        # Only 1 command per hook will be run
+        unix:
+          filteredCommands:
+            # (required) A run command is a command to run in order to obtain some information about the Host OS and OS version
+            - getHostOSInfoCommand: uname
```

Comment:
The idea with the hooks is that they are configured by the addon developer and used to extend minikube in a way that the current ecosystem cannot do with kubernetes configurations alone. Having to separately configure a plugin for behavior which could be done automatically would not provide a good user experience as it gives an opportunity to misconfigure something.

Reply:
I don't mind a different format for the config file as long as we can achieve the goal. My proposal would allow the plugin developer flexibility for doing different things on different types of hosts which could also be done in a script outside of minikube. I believe your proposal is simpler and better.